modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyListProperty.java

```diff
@@ -114,14 +114,16 @@ public boolean equals(Object obj) {
         if (!(obj instanceof List)) {
             return false;
         }
-        final List list = (List)obj;
+
+        @SuppressWarnings("unchecked")
+        final List<E> list = (List<E>)obj;  // safe cast as elements are only referenced
```

Comment:
It's safe as you're never actually using the type `E`, but I suppose the 2nd list can also be `List<?>` as only `equals` is called on it.  The 2nd `ListIterator` would then also be `ListIterator<?>`.

Casting to a specific generic type (like `List<E>`) never verifies if all the elements are of that type, you're just telling the compiler that they are.

Only when you do something like:

    E e = list.get(0);

... you may get a CCE here (but the compiler allowed it because you casted it to `List<E>` earlier).  That's why I say it is safe because the follow up code does not do any such action (it in fact does `Object o2 = e2.next()` -- it would be bad if it was `E o2 = e2.next`).

Looking at it again, I think making the 2nd `List` a `List<?>` is better, as we don't know yet at that stage that they all are of type `E` (even though it doesn't matter for the code following it).

Reply:
It also removes the need for the `SuppressWarnings` here.

Reply:
I suggest the following method:

```java
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof List<?> list)) {
            return false;
        }

        if (size() != list.size()) {
            return false;
        }

        ListIterator<E> e1 = listIterator();
        ListIterator<?> e2 = list.listIterator();
        while (e1.hasNext() && e2.hasNext()) {
            if (!Objects.equals(e1.next(), e2.next())) {
                return false;
            }
        }
        return true;
    }
```

Reply:
There is a lot to not like about this method.  Especially because apparently you can pass in a `List` to this equals method that is for a `ReadOnlyListProperty`.  How is a list equal to a property?  What sense does it make to do this:

     ReadOnlyListProperty<String> x;

     x.equals(List.of("a", "b", "c"));

Shouldn't that be:

     x.get().equals(List.of("a", "b", "c"));

?

Reply:
> So... it's also a list.

I think it's a big mess, you can already see that in the hierarchy, extending both `ObservableObjectValue<ObservableList<E>>` (which is `ObservableValue<T>`) but also being a `ObservableList<E>`.  One or the other is going to break; either it is a property containing a list, or it is a list, it can't be both.

Either I can do:

     listProperty.equals(normalList);

or:

     listProperty.equals(aNormalPropertyContainingAList);

It looks like the first was chosen (I suppose because the `List` interface defines `equals` in that way) and the 2nd case will break (and is unfixable as equals needs to be symmetric).

Looking a bit further, I think extending `ObservableObjectValue<ObservableList<E>>` was a mistake here. It's also what contributes to making the collection properties so confusing (ie. you can replace all elements, or you can replace the entire list, which are two different actions that require two different forms of listeners) -- no other properties allow you to replace it's "container" -- ie. if properties had used an `AtomicReference` internally to store their value, they would not offer me the option to replace either the entire `AtomicReference` or only the value of the reference.

Reply:
I've adjusted it now to be a bit more modern Java as per your suggestion.

Reply:
I agree that this is dubious design, and having worked a bit on observable collections I also found some questionable decisions with regards to the type hierarchy. For example. `ListExpression` and some subclasses should have been interfaces, all the way down to `ListPropertyBase` where a state is actually present. As for the equals, I wonder if it should have been defined in `ListExpression` instead, though that would bring the question about bindings - what makes a binding equal to another.

Right now, what we're doing right now is copying the `AbstractList` equals implementation essentially. I think it's fine, but points, again, to a possible hierarchy problem.

On another note, I'm looking at `ListExpression`'s `add` and other modifying methods. It's calling `EMPTY_LIST.add(element)` where `ObservableList EMPTY_LIST = FXCollections.emptyObservableList();`. Wouldn't that throw an exception since it's an empty unmodifiable observable list?

Reply:
> On another note, I'm looking at ListExpression's add and other modifying methods. It's calling EMPTY_LIST.add(element) where ObservableList EMPTY_LIST = FXCollections.emptyObservableList();. Wouldn't that throw an exception since it's an empty unmodifiable observable list?

I saw that code and its intentional.  I think the java doc at the top gives the hint:

> If the wrapped list of a {@code ListExpression} is {@code null}, all methods implementing the {@code List} interface will behave as if they were applied to an immutable empty list.